File: 0338-counting-bits/0338-counting-bits.go
func countBits(n int) []int {
    ans := []int{}
    for i := 0; i <= n; i++ {
        ans = append(ans, count(i))
    }

    return ans
}

func count(n int) int {
    if n < 2 {
        return n
    }
    
    total := 0
    for n > 0 {
        total += n % 2
        n /= 2
    }

    return total
}

File: 0338-counting-bits/0338-counting-bits.cpp
/*
0  -> 0000 -> 0
2  -> 0010 -> 1
4  -> 0100 -> 1
6  -> 0110 -> 2
8  -> 1000 -> 1
13 -> 1101 -> 3
*/

class Solution {
public:
    vector<int> countBits(int n) {
        vector<int> ans(n + 1);
        for (int i = 0; i <= n; i++) {
            ans[i] = ans[i / 2] + i % 2;
        }

        return ans;
    }
};